packages/react/src/components/Checkbox/Checkbox.tsx

```diff
@@ -107,30 +116,32 @@ class Checkbox extends AutoControlledComponent<WithAsProp<CheckboxProps>, Checkb
   }
 
   getInitialAutoControlledState(): CheckboxState {
-    return { checked: false }
+    return { checked: false, indeterminate: false }
   }
 
   handleChange = (e: React.ChangeEvent) => {
     // Checkbox component doesn't present any `input` component in markup, however all of our
     // components should handle events transparently.
-    const { disabled } = this.props
+    const { disabled, indeterminate } = this.props
     const checked = !this.state.checked
+    const isIndeterminate = !this.state.indeterminate
 
     if (!disabled) {
-      this.setState({ checked })
-      _.invoke(this.props, 'onChange', e, { ...this.props, checked })
+      this.setState({ checked, indeterminate })
+      _.invoke(this.props, 'onChange', e, { ...this.props, checked, isIndeterminate })
     }
   }
 
   handleClick = (e: React.MouseEvent | React.KeyboardEvent) => {
-    const { disabled } = this.props
+    const { disabled, indeterminate } = this.props
     const checked = !this.state.checked
+    const isIndeterminate = !this.state.indeterminate
 
     if (!disabled) {
-      this.setState({ checked })
+      this.setState({ checked, indeterminate })
 
-      _.invoke(this.props, 'onClick', e, { ...this.props, checked })
-      _.invoke(this.props, 'onChange', e, { ...this.props, checked })
+      _.invoke(this.props, 'onClick', e, { ...this.props, checked, isIndeterminate })
+      _.invoke(this.props, 'onChange', e, { ...this.props, checked, isIndeterminate })
```

Comment:
```suggestion
      _.invoke(this.props, 'onChange', e, { ...this.props, checked, indeterminate: !this.state.indeterminate })
```

We use that the same signature as props